x-pack/plugins/triggers_actions_ui/public/application/components/builtin_action_types/es_index/es_index_params.tsx

```diff
@@ -20,38 +34,152 @@ export const IndexParamsFields = ({
   editAction,
   messageVariables,
   errors,
+  actionConnector,
 }: ActionParamsProps<IndexActionParams>) => {
   const { docLinks } = useKibana().services;
-  const { documents } = actionParams;
+  const { documents, indexOverride } = actionParams;
+
+  const defaultAlertHistoryIndexSuffix = AlertHistoryDefaultIndexName.replace(
+    ALERT_HISTORY_PREFIX,
+    ''
+  );
+
+  const getDocumentToIndex = (doc: Array<Record<string, any>> | undefined) =>
+    doc && doc.length > 0 ? ((doc[0] as unknown) as string) : undefined;
+
+  const [documentToIndex, setDocumentToIndex] = useState<string | undefined>(
+    getDocumentToIndex(documents)
+  );
+  const [alertHistoryIndexSuffix, setAlertHistoryIndexSuffix] = useState<string>(
+    indexOverride ? indexOverride.replace(ALERT_HISTORY_PREFIX, '') : defaultAlertHistoryIndexSuffix
+  );
+  const [usePreconfiguredSchema, setUsePreconfiguredSchema] = useState<boolean>(false);
+
+  useEffect(() => {
+    setDocumentToIndex(getDocumentToIndex(documents));
+  }, [documents]);
+
+  useEffect(() => {
+    if (actionConnector?.id === AlertHistoryEsIndexConnectorId) {
+      setUsePreconfiguredSchema(true);
+      editAction('documents', [JSON.stringify(AlertHistoryDocumentSchema)], index);
+      setDocumentToIndex(JSON.stringify(AlertHistoryDocumentSchema));
+    } else {
+      setUsePreconfiguredSchema(false);
+      editAction('documents', undefined, index);
+      setDocumentToIndex(undefined);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [actionConnector?.id]);
 
   const onDocumentsChange = (updatedDocuments: string) => {
     try {
       const documentsJSON = JSON.parse(updatedDocuments);
       editAction('documents', [documentsJSON], index);
+      setDocumentToIndex(updatedDocuments);
     } catch (e) {
       // set document as empty to turn on the validation for non empty valid JSON object
       editAction('documents', [{}], index);
+      setDocumentToIndex(undefined);
     }
   };
 
-  return (
+  const documentsFieldLabel = i18n.translate(
+    'xpack.triggersActionsUI.components.builtinActionTypes.indexAction.documentsFieldLabel',
+    {
+      defaultMessage: 'Document to index',
```

Comment:
This wasn't clear to me the first time I read it.  Is there something more specific you can use?